// vue-use-fixed-header/dist/index.js
var K = Object.defineProperty, Q = Object.defineProperties;
var W = Object.getOwnPropertyDescriptors;
var N = Object.getOwnPropertySymbols;
var X = Object.prototype.hasOwnProperty, Z = Object.prototype.propertyIsEnumerable;
var I = (t, n, i) => n in t ? K(t, n, { enumerable: !0, configurable: !0, writable: !0, value: i }) : t[n] = i, u = (t, n) => {
  for (var i in n || (n = {}))
    X.call(n, i) && I(t, i, n[i]);
  if (N)
    for (var i of N(n))
      Z.call(n, i) && I(t, i, n[i]);
  return t;
}, L = (t, n) => Q(t, W(n));
import { ref as C, onMounted as q, onBeforeUnmount as ee, shallowRef as ne, watch as te, unref as r, readonly as oe, computed as F } from "vue";
const y = typeof window != "undefined";
function ie() {
  const t = C(!1);
  if (!y)
    return t;
  const n = window.matchMedia("(prefers-reduced-motion: reduce)"), i = () => t.value = n.matches;
  return q(() => {
    var l;
    i(), (l = n.addEventListener) == null || l.call(n, "change", i);
  }), ee(() => {
    var l;
    (l = n.removeEventListener) == null || l.call(n, "change", i);
  }), t;
}
const B = "cubic-bezier(0.16, 1, 0.3, 1)", re = {
  enterStyles: {
    transition: `all 0.35s ${B} 0s`,
    transform: "translateY(0px)"
  },
  leaveStyles: {
    transition: `all 0.5s ${B} 0s`,
    transform: "translateY(-101%)"
  }
};
function le(t, n = {}) {
  const { enterStyles: i, leaveStyles: l } = re, P = () => n.root || n.root === null ? r(n.root) : null, b = () => n.transitionOpacity === void 0 ? !1 : r(n.transitionOpacity), Y = ie(), s = {
    resizeObserver: void 0,
    initResizeObserver: !1,
    isListeningScroll: !1,
    isHovering: !1
  }, v = ne({}), a = C(
    0
    /* READY */
  ), m = (e) => v.value = e, g = () => v.value = {}, h = (e) => a.value = e;
  function d() {
    const e = P();
    return e != null ? e : document.documentElement;
  }
  function p() {
    const e = d();
    return e ? e.scrollTop : 0;
  }
  function _() {
    const e = r(t);
    if (!e)
      return !1;
    const { position: o, display: c } = window.getComputedStyle(e);
    return (o === "fixed" || o === "sticky") && c !== "none";
  }
  function A() {
    const e = r(t);
    if (!e)
      return 0;
    let o = e.scrollHeight;
    const { marginTop: c, marginBottom: f } = window.getComputedStyle(e);
    return o += Number.parseFloat(c) + Number.parseFloat(f), o;
  }
  function U() {
    s.resizeObserver = new ResizeObserver(() => {
      if (!s.initResizeObserver)
        return s.initResizeObserver = !0;
      z();
    });
    const e = d();
    e && s.resizeObserver.observe(e);
  }
  function R() {
    a.value !== 1 && (E(), m(L(u(u({}, i), b() ? { opacity: 1 } : {}), {
      visibility: ""
    })), h(
      1
      /* ENTER */
    ));
  }
  function V() {
    a.value !== 2 && (m(u(u({}, l), b() ? { opacity: 0 } : {})), h(
      2
      /* LEAVE */
    ), $());
  }
  function w(e) {
    if (E(), !r(t) || e.target !== r(t) || e.propertyName !== "transform")
      return;
    const { transform: o } = window.getComputedStyle(r(t));
    o !== "matrix(1, 0, 0, 1, 0, 0)" && m(L(u({}, l), {
      visibility: "hidden"
    }));
  }
  function $() {
    const e = r(t);
    e && e.addEventListener("transitionend", w);
  }
  function E() {
    const e = r(t);
    e && e.removeEventListener("transitionend", w);
  }
  function k() {
    let e = y ? p() : 0;
    return () => {
      const o = p(), c = o <= A(), f = o < e, S = o > e, x = Math.abs(o - e);
      if (c)
        return R();
      x < 10 || (s.isHovering || (f ? R() : S && V()), e = p());
    };
  }
  const O = k();
  function D() {
    const e = d();
    if (!e)
      return;
    (e === document.documentElement ? document : e).addEventListener("scroll", O, { passive: !0 }), s.isListeningScroll = !0;
  }
  function G() {
    const e = d();
    if (!e)
      return;
    (e === document.documentElement ? document : e).removeEventListener("scroll", O), s.isListeningScroll = !1;
  }
  function T(e) {
    var o, c;
    s.isHovering = (c = (o = r(t)) == null ? void 0 : o.contains(e.target)) != null ? c : !1;
  }
  function j() {
    document.addEventListener("pointermove", T);
  }
  function J() {
    document.removeEventListener("pointermove", T);
  }
  function z() {
    const e = _();
    s.isListeningScroll ? e || (H(), g()) : e && (D(), j());
  }
  function H() {
    G(), J();
  }
  return y && te(
    () => [r(t), d(), Y.value, r(n.watch)],
    ([e, o, c], f, S) => {
      !c && e && (o || o === null) && (U(), z()), S(() => {
        var M;
        H(), g(), (M = s.resizeObserver) == null || M.disconnect(), s.initResizeObserver = !1;
      });
    },
    { immediate: !0, flush: "post" }
  ), {
    styles: oe(v),
    isLeave: F(
      () => a.value === 2
      /* LEAVE */
    ),
    isEnter: F(
      () => a.value === 1
      /* ENTER */
    )
  };
}
export {
  le as useFixedHeader
};
